libsweep/src/arch/win/serial_win.cc

```diff
@@ -0,0 +1,280 @@
+  COMMTIMEOUTS timeouts;
+  if (!GetCommTimeouts(hComm, &timeouts)) {
+	  *error = error_construct("retrieving current serial port timeouts failed");
+	  CloseHandle(hComm);
+	  return nullptr;
+  }
+  timeouts.ReadIntervalTimeout = 50;         // max time between arrival of two bytes before ReadFile() returns
+  timeouts.ReadTotalTimeoutConstant = 50;    // used to calculate total time-out period for read operations
+  timeouts.ReadTotalTimeoutMultiplier = 10;  // used to calculate total time-out period for read operations
+  timeouts.WriteTotalTimeoutConstant = 50;   // used to calculate total time-out period for write operations
+  timeouts.WriteTotalTimeoutMultiplier = 10; // used to calculate total time-out period for write operations
+
+  // set the timeouts
+  if (!SetCommTimeouts(hComm, &timeouts)) {
+    *error = error_construct("setting serial port timeouts failed");
+    CloseHandle(hComm);
+    return nullptr;
+  }
+
+  // create the overlapped os reader
+  OVERLAPPED osReader = {0};
+  OVERLAPPED osWrite = {0};
+
+  // Create the overlapped read event. Must be closed before exiting
+  osReader.hEvent = CreateEvent(NULL, TRUE, FALSE, NULL);
+  if (osReader.hEvent == NULL) {
+    *error = error_construct("creating overlapped read event failed");
+    CloseHandle(hComm);
+    return nullptr;
+  }
+  // Create the overlapped write event. Must be closed before exiting
+  osWrite.hEvent = CreateEvent(NULL, TRUE, FALSE, NULL);
+  if (osWrite.hEvent == NULL) {
+    *error = error_construct("creating overlapped write event failed");
+    CloseHandle(hComm);
+    return nullptr;
+  }
+
+  // set the comm mask
+  if (!SetCommMask(hComm, EV_RXCHAR | EV_ERR)) {
+    *error = error_construct("setting comm mask failed");
+    CloseHandle(hComm);
+    CloseHandle(osReader.hEvent);
+    CloseHandle(osWrite.hEvent);
+    return nullptr;
+  }
+
+  // purge the comm port of any pre-existing data or errors
+  if (!PurgeComm(hComm, PURGE_RXABORT | PURGE_TXABORT | PURGE_RXCLEAR | PURGE_TXCLEAR)) {
+    *error = error_construct("flushing serial port failed during serial device construction");
+    CloseHandle(hComm);
+    CloseHandle(osReader.hEvent);
+    CloseHandle(osWrite.hEvent);
+    return nullptr;
+  }
+
+  // create the serial device
+  auto out = new device{hComm, osReader, osWrite};
+
+  return out;
+}
+
+void device_destruct(device_s serial) {
+  SWEEP_ASSERT(serial);
+
+  error_s ignore = nullptr;
+  device_flush(serial, &ignore);
+
+  // try to close the serial port
+  if (!CloseHandle(serial->fd)) {
+    ignore = error_construct("closing serial port failed");
+  }
+
+  // try to close the overlapped read event
+  if (!CloseHandle(serial->ro_.hEvent)) {
+    ignore = error_construct("closing serial port overlapped read event failed");
+  }
+
+  // try to close the overlapped write event
+  if (!CloseHandle(serial->wo_.hEvent)) {
+    ignore = error_construct("closing serial port overlapped write event failed");
```

Comment:
Hm no reason to set errors here we are not exposing anyway. You can just call your functions and hope for the best.